### ISSUE TITLE
General questions about linalgcpp

### ISSUE THREAD

**Issue by chakshinglee** [gelever/linalgcpp#23]
https://github.com/gelever/linalgcpp/blob/6186c0978e517b7c3738eb9c30ddc53987a20666/include/vectorview.hpp#L21-L31

How do to determine if a view is owned? 

By the way, should "most" be "must" in the documentation?

**Comment by gelever** [comment]
Perhaps this documentation is poorly worded.  You can only create a copy of a view if the original view is not const.  This prevents modifying a const view by simply copying it into a new non-const view (but the data they both view is the same).

Yes, good catch on the typo.

**Comment by gelever** [comment]
Yes, that would definitely be an option for vector operator equal. I think that would imply that `lhs` and `rhs` are pointing to the same data. Since the chances of them having the exact floating point entries is pretty slim.

**Comment by chakshinglee** [comment]
(Apparently I cannot reply to a comment right below it?)

For the ownership of `VectorView`, I still don't understand. So `VectorView` by definition does not own the data, right? The ownership here means `const`?

**Comment by chakshinglee** [comment]
https://github.com/gelever/linalgcpp/blob/6186c0978e517b7c3738eb9c30ddc53987a20666/include/vectorview.hpp#L247-L257

Why the move assignment needs to copy the data? It is not safe the move the data?

**Comment by gelever** [comment]
Correct, `VectorView` does not own any data.

Operators are defined on the underlying data, not the view itself.  I originally had it the other way around.  In practice, this is much more useful.  Because of the change, ownership now is not really an issue.

**Comment by chakshinglee** [comment]
https://github.com/gelever/linalgcpp/blob/6186c0978e517b7c3738eb9c30ddc53987a20666/include/vectorview.hpp#L476-L481

I found this function counter-intuitive. I understand what the function is doing, just to me mathematically there are two natural multiplications for vectors, one is vector * scalar = vector, one is vector * vector = scalar (of course, this is just my own intuition).  I totally agree this is a useful function, I just dont find vector1 *= vector2 easy to follow.

**Comment by chakshinglee** [comment]
If by definition `VectorView` does not own data, then what is the meaning of "if you own the view" in the documentation?

**Comment by gelever** [comment]
Ownership refer(ed) to the `VectorView` itself. I will update the docs to clarify the restrictions on `VectorView`

**Comment by chakshinglee** [comment]
https://github.com/gelever/linalgcpp/blob/6186c0978e517b7c3738eb9c30ddc53987a20666/include/sparsematrix.hpp#L344-L347

For a general linear algebra library, this function does not make much sense to me (although I see how it can be used in smoothG). And in general if `values` is a non-diagonal matrix (with nnz = number of rows), the function will not complain, so it can be hard to debug if a non-diagonal matrix is passed in accidentally. 

The same comment applies to three other similar functions.

**Comment by gelever** [comment]
For scaling, I think those are useful enough to belong in the linear algebra library.  I agree that scaling by a matrix may be dangerous and requires either more clear documentation or stricter debug checks.

Yes, those deep copy for now, until I create `BlockVectorView`.  It is a temp solution until I get there.

**Comment by chakshinglee** [comment]
I agree the functionality of the function itself is useful. I just cannot see what does it mean by scaling rows of a matrix by another matrix (thinking from the point of reading some code that uses this function).

By the way, how do you think about making a special matrix class for diagonal matrices? This needs not be in linalgcpp (as the class is probably too restrictive in general to be used), but can be created in smoothG (derived from `Operator` or something).

**Comment by chakshinglee** [comment]
https://github.com/gelever/linalgcpp/blob/6186c0978e517b7c3738eb9c30ddc53987a20666/include/sparsematrix.hpp#L342

MaxNorm is not describing what it actually computes? This wiki page https://en.wikipedia.org/wiki/Matrix_norm calls it "entrywise" L1 norm (which is different from the induced L1 norm). Of course, that's just a reference.

**Comment by gelever** [comment]
Ah ok, I thought you meant scaling matrices in general.  Yes, for scaling by matrix, the function name is misleading.

For matrix norm, you are correct. It is incorrectly named.  I needed something that would determine if a matrix is close enough to 0 and named it w/o looking up the proper name.  I'll modify it.
